Add terraformtype Support for github.com/hashicorp/terraform-plugin-sdk/helper/schema.CustomizeDiffFunc

Version 2 of the Terraform Plugin SDK changes its type signature:

```
aws/resource_aws_api_gateway_authorizer.go:23:3: cannot use resourceAwsApiGatewayAuthorizerCustomizeDiff (type func(*schema.ResourceDiff, interface {}) error) as type schema.CustomizeDiffFunc in field value
```

To support downstream Analyzers wishing to handle this, we can add constants and helper functions in a `helper/terraformtype/helper/schema` package similar to the others.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,9 +11,12 @@ ENHANCEMENTS
 * helper/analysisutils: `DeprecatedWithReplacementSelectorExprRunner` now implements `SuggestedFixes`
 * helper/astutils: Add `HasFieldListLength`, `IsExprTypeBool`, `IsExprTypeMapStringInterface` functions
 * helper/terraformtype/helper/customdiff: Constants and functions for working with entire package
+* helper/terraformtype/helper/schema: Constants and functions for working with `CustomizeDiffFunc` type
 * helper/terraformtype/helper/schema: Constants and functions for working with `StateUpgradeFunc` type
 * helper/terraformtype/terraform: Constants and functions for working with `ResourceProvider` type
 * passes/helper/customdiff: Passes for collecting `customdiff` function information for further analysis
+* passes/helper/schema: Pass for collecting `CustomizeDiffFunc` function information for further analysis
+* passes/helper/schema: Pass for collecting `StateUpgradeFunc` function information for further analysis
 * passes/R007: Support suggested fix
 * passes/R008: Support suggested fix
 * passes/V002: Support suggested fix

--- a/helper/terraformtype/helper/schema/type_customizedifffunc.go
+++ b/helper/terraformtype/helper/schema/type_customizedifffunc.go
@@ -1,0 +1,87 @@
+package schema
+
+import (
+	"go/ast"
+	"go/token"
+	"go/types"
+
+	"github.com/bflad/tfproviderlint/helper/astutils"
+)
+
+const (
+	TypeNameCustomizeDiffFunc = `CustomizeDiffFunc`
+)
+
+// IsFuncTypeCustomizeDiffFunc returns true if the FuncType matches expected parameters and results types
+func IsFuncTypeCustomizeDiffFunc(node ast.Node, info *types.Info) bool {
+	funcType := astutils.FuncTypeFromNode(node)
+
+	if funcType == nil {
+		return false
+	}
+
+	if !astutils.HasFieldListLength(funcType.Params, 2) {
+		return false
+	}
+
+	if !astutils.IsFieldListTypePackageType(funcType.Params, 0, info, PackagePath, TypeNameResourceDiff) {
+		return false
+	}
+
+	if !astutils.IsFieldListType(funcType.Params, 1, astutils.IsExprTypeInterface) {
+		return false
+	}
+
+	if !astutils.HasFieldListLength(funcType.Results, 1) {
+		return false
+	}
+
+	return astutils.IsFieldListType(funcType.Results, 0, astutils.IsExprTypeError)
+}
+
+// IsTypeCustomizeDiffFunc returns if the type is CustomizeDiffFunc from the customdiff package
+func IsTypeCustomizeDiffFunc(t types.Type) bool {
+	switch t := t.(type) {
+	case *types.Named:
+		return IsNamedType(t, TypeNameCustomizeDiffFunc)
+	case *types.Pointer:
+		return IsTypeCustomizeDiffFunc(t.Elem())
+	default:
+		return false
+	}
+}
+
+// CustomizeDiffFuncInfo represents all gathered CustomizeDiffFunc data for easier access
+type CustomizeDiffFuncInfo struct {
+	AstFuncDecl *ast.FuncDecl
+	AstFuncLit  *ast.FuncLit
+	Body        *ast.BlockStmt
+	Node        ast.Node
+	Pos         token.Pos
+	Type        *ast.FuncType
+	TypesInfo   *types.Info
+}
+
+// NewCustomizeDiffFuncInfo instantiates a CustomizeDiffFuncInfo
+func NewCustomizeDiffFuncInfo(node ast.Node, info *types.Info) *CustomizeDiffFuncInfo {
+	result := &CustomizeDiffFuncInfo{
+		TypesInfo: info,
+	}
+
+	switch node := node.(type) {
+	case *ast.FuncDecl:
+		result.AstFuncDecl = node
+		result.Body = node.Body
+		result.Node = node
+		result.Pos = node.Pos()
+		result.Type = node.Type
+	case *ast.FuncLit:
+		result.AstFuncLit = node
+		result.Body = node.Body
+		result.Node = node
+		result.Pos = node.Pos()
+		result.Type = node.Type
+	}
+
+	return result
+}

--- a/helper/terraformtype/helper/schema/type_customizedifffunc_test.go
+++ b/helper/terraformtype/helper/schema/type_customizedifffunc_test.go
@@ -1,0 +1,186 @@
+package schema
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"testing"
+)
+
+func TestIsFuncTypeCustomizeDiffFunc(t *testing.T) {
+	boolIdent := &ast.Ident{
+		Name: types.Typ[types.Bool].String(),
+	}
+	errorIdent := &ast.Ident{
+		Name: "error",
+	}
+	interfaceAst := &ast.InterfaceType{
+		Methods: &ast.FieldList{},
+	}
+	packageIdent := &ast.Ident{
+		Name: PackageName,
+	}
+	packageType := types.NewPackage(PackagePath, PackageName)
+	packageNameType := types.NewPkgName(token.NoPos, packageType, packageType.Name(), packageType)
+	packageFuncIdent := &ast.Ident{
+		Name: TypeNameResourceDiff,
+	}
+	packageSelectorExpr := &ast.SelectorExpr{
+		Sel: packageFuncIdent,
+		X:   packageIdent,
+	}
+	typesInfo := &types.Info{
+		Types: map[ast.Expr]types.TypeAndValue{
+			boolIdent: types.TypeAndValue{
+				Type: types.Typ[types.Bool],
+			},
+			errorIdent: types.TypeAndValue{
+				Type: types.NewNamed(types.NewTypeName(token.NoPos, nil, "error", nil), nil, nil),
+			},
+			interfaceAst: types.TypeAndValue{
+				Type: types.NewInterfaceType(nil, nil),
+			},
+			packageSelectorExpr: types.TypeAndValue{
+				Type: types.NewNamed(types.NewTypeName(token.NoPos, packageType, TypeNameResourceDiff, nil), nil, nil),
+			},
+		},
+		Uses: map[*ast.Ident]types.Object{
+			packageIdent: packageNameType,
+		},
+	}
+
+	testCases := []struct {
+		Name     string
+		Node     ast.Node
+		Info     *types.Info
+		Expected bool
+	}{
+		{
+			Name: fmt.Sprintf("func(*%s.%s, interface{}) bool", PackagePath, TypeNameResourceDiff),
+			Node: &ast.FuncLit{
+				Type: &ast.FuncType{
+					Params: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: &ast.StarExpr{
+									X: packageSelectorExpr,
+								},
+							},
+							{
+								Type: interfaceAst,
+							},
+						},
+					},
+					Results: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: boolIdent,
+							},
+						},
+					},
+				},
+			},
+			Info:     typesInfo,
+			Expected: false,
+		},
+		{
+			Name: fmt.Sprintf("func(*%s.%s, interface{}) error", PackagePath, TypeNameResourceDiff),
+			Node: &ast.FuncLit{
+				Type: &ast.FuncType{
+					Params: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: &ast.StarExpr{
+									X: packageSelectorExpr,
+								},
+							},
+							{
+								Type: interfaceAst,
+							},
+						},
+					},
+					Results: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: errorIdent,
+							},
+						},
+					},
+				},
+			},
+			Info:     typesInfo,
+			Expected: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			got := IsFuncTypeCustomizeDiffFunc(testCase.Node, testCase.Info)
+
+			if got != testCase.Expected {
+				t.Errorf("got %t, expected %t", got, testCase.Expected)
+			}
+		})
+	}
+}
+
+func TestIsTypeCustomizeDiffFunc(t *testing.T) {
+	testCases := []struct {
+		Name      string
+		InputType types.Type
+		Expected  bool
+	}{
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, TypeNameCustomizeDiffFunc),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameCustomizeDiffFunc, nil),
+				nil,
+				nil,
+			),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("*%s.%s", PackagePath, TypeNameCustomizeDiffFunc),
+			InputType: types.NewPointer(types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameCustomizeDiffFunc, nil),
+				nil,
+				nil,
+			)),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, "Not"),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), "Not", nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", "incorrect/path", TypeNameCustomizeDiffFunc),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage("incorrect/path", "path"), TypeNameCustomizeDiffFunc, nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name:      "string",
+			InputType: types.Typ[types.String],
+			Expected:  false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			got := IsTypeCustomizeDiffFunc(testCase.InputType)
+
+			if got != testCase.Expected {
+				t.Errorf("got %t, expected %t", got, testCase.Expected)
+			}
+		})
+	}
+}

--- a/passes/helper/schema/customizedifffuncinfo/customizedifffuncinfo.go
+++ b/passes/helper/schema/customizedifffuncinfo/customizedifffuncinfo.go
@@ -1,0 +1,40 @@
+package customizedifffuncinfo
+
+import (
+	"go/ast"
+	"reflect"
+
+	"github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "customizedifffuncinfo",
+	Doc:  "find github.com/hashicorp/terraform-plugin-sdk/helper/schema CustomizeDiffFunc declarations for later passes",
+	Requires: []*analysis.Analyzer{
+		inspect.Analyzer,
+	},
+	Run:        run,
+	ResultType: reflect.TypeOf([]*schema.CustomizeDiffFuncInfo{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	nodeFilter := []ast.Node{
+		(*ast.FuncDecl)(nil),
+		(*ast.FuncLit)(nil),
+	}
+	var result []*schema.CustomizeDiffFuncInfo
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		if !schema.IsFuncTypeCustomizeDiffFunc(n, pass.TypesInfo) {
+			return
+		}
+
+		result = append(result, schema.NewCustomizeDiffFuncInfo(n, pass.TypesInfo))
+	})
+
+	return result, nil
+}

--- a/passes/helper/schema/customizedifffuncinfo/customizedifffuncinfo_test.go
+++ b/passes/helper/schema/customizedifffuncinfo/customizedifffuncinfo_test.go
@@ -1,0 +1,15 @@
+package customizedifffuncinfo
+
+import (
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+func TestValidateAnalyzer(t *testing.T) {
+	err := analysis.Validate([]*analysis.Analyzer{Analyzer})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}

